app/api/git-query.js

```diff
@@ -1,36 +1,81 @@
+const EasyZip = require('easy-zip').EasyZip;
 const util = require('util');
-
 const stat = util.promisify(fs.stat);
 const mkdir = util.promisify(fs.mkdir);
 
-let basePath = __dirname + '/repositories/';
+const authenticate      = require('./authentication.js');
+const express           = require('express');
+const config            = require('./config.js');
+const signedUrl         = require('./signed-url.js');
 
+const basePath = __dirname + '/repositories/';
 
-module.exports = function gitQuery(options){
 
-    return async function get(req, res, next){
+function gitQuery(options){
+    
+    let router = express.Router({mergeParams:true});
 
-        let q = req.query;
-        
-        let repositoryName = req.params.repository;
-        let files = await getUpdatesFiles(repositoryName, q.branch, q.date, q.ignoreFiles, q.allowedExtenstions);
+    router.get   ('/',             authenticate, authorized, get); 
+    router.get   ('/signed-url',   authenticate, authorized, signedUrl.get);
+    router.get   ('/download',     signedUrl.isValid, download);
 
-        if(q.zip){
-            let compressedFiles = await zipFile(repositoryName, files, q.branch);
-            let zip4 = new EasyZip();
-            return zip4.batchAdd(compressedFiles, function(){
-                return zip4.writeToResponse(res,`${q.branch}`);
-            });
+    return router;
+
+    async function get(req, res){
+        try{
+            let q = req.query;
+            
+            let repositoryName = req.params.repository;
+            let files = await getUpdatesFiles(repositoryName, q.branch, q.date, q.ignoreFiles, q.allowedExtenstions);
+
+            res.status(200).send(files);
         }
-        res.status(200).send(files);
+        catch(err) {
+            res.status(500).send('Unknown error occurred');
+        };
     }
 
+    async function download(req, res){
+        try{
+            
+            let q = req.decryptedInfo;
+            
+            let repositoryName = req.params.repository;
+            let files = await getUpdatesFiles(repositoryName, q.branch, q.date, q.ignoreFiles, q.allowedExtenstions);
+
+            if(files && files.length > 0){
+                let compressedFiles = await zipFile(repositoryName, files, q.branch);
+                let zip4 = new EasyZip();
+                return zip4.batchAdd(compressedFiles, function(){
+                    return zip4.writeToResponse(res,`${q.branch}`);
+                });    
+            }
+            res.status(200).send('No files found for download');
+        }
+        catch(err) {
+            winston.error(err)
+            res.status(500).send('Unknown error occurred');
+        };
+    }
+
+    async function authorized(req, res, next){
+        try{
+
+            if(!req.user || !authenticate.isInRole(req.user, ['Administrator', 'oasisArticleEditor'])){
+                return res.status(401).send('You are not authorized to access this resource');
```

Comment:
403 ~not 401~

https://stackoverflow.com/questions/3297048/403-forbidden-vs-401-unauthorized-http-responses